core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -768,6 +775,28 @@ public PlanOptimizers(
                             // Must run before AddExchanges and after ReplicateSemiJoinInDelete
                             // to avoid temporarily having an invalid plan
                             new DetermineSemiJoinDistributionType(costComparator, taskCountEstimator)))));
+
+            // Must run before DetermineTableScanNodePartitioning because PushJoinIntoTableScan creates new table handles
+            // without useConnectorNodePartitioning set which will cause planner to fail
+            builder.add(new IterativeOptimizer(
+                    ruleStats,
+                    statsCalculator,
+                    estimatedExchangesCostCalculator,
+                    pushIntoTableScanRules));
+
+            // DetermineTableScanNodePartitioning needs to ensure all table handles have proper partitioning determined
+            builder.add(new IterativeOptimizer(
+                    ruleStats,
+                    statsCalculator,
+                    costCalculator,
+                    ImmutableSet.of(
+                            // Must run before AddExchanges
+                            new PushDeleteIntoConnector(metadata),
```

Comment:
What's the point here? Looks like a copy. Can we omit `PushDeleteIntoConnector`?

Reply:
It is indeed copied. I'm unsure if the rule @sopel39 talked about depends on the other rules in any way.

Looks like both `OptimizeDuplicateInsensitiveJoins` and `PushDeleteIntoConnector` can be removed safely?

Reply:
You can also put the copied IOp into variable and add it twice, to avoid copy-paste huge block of text

Reply:
The other rules don't need to be re-run since their patterns can't be matched anymore and `DetermineTableScanNodePartitioning` doesn't seem to be depend on those.

Removed. cc: @sopel39